### DISCUSSION TITLE
support template file loaders

### DISCUSSION THREAD

**Discussion by nerdoc** [tetra-framework/tetra#23] [Ideas]
While I really like the idea having templates in visual context next to the code as inline strings, this has a few drawbacks too:

* You need to have an editor that supports highlighting - without that a template is badly readable. E.g. PyCharm does not support this.
* As long as templates are tiny, this is cool. When they get larger, you won't feel good as you loose the bigger view you gained with the advantage of code-template proximity. This also myabe is an editor issue, as e.g. code folding is not possible there.

So I would suggest to at least optionally add template loading from file, like:

```python
class MyComponent(Component):
    ...
    template: django_html = get_template("my_app/components/my_component.html)
```

Maybe it could be a default too: if you don't specify a `template` attribute (with a str), the component looks in the default folder (app/template/app/components e.g.).

I can understand if you'd like to stay pure here too, and as I said in the first place, I support the idea of having inline templates. But it would be helpful to at least *support* external templates too.

**Comment by samwillis** [comment]
Thats a great idea, and thanks for the other comments! I have been pretty tied up with contract work but will take a proper look at this and your issues maybe this weekend.

The general feedback I have had is that the inline templates are good but maybe one step too far to encourage adoption. Tetra is the result of having *a lot* of ideas.

**Reply by nerdoc** [reply]
Yea, I saw that. But that makes it promising, as it's fresh, as Vue.js was. I'm afraid I can't help much with knowledge, as I don't know much of the internals, but I REALLY tried all other projects, and all had major flaws or are not suitable for larger projects, as they are not component based (HTMX, Turbo-Django), have a worse API than Tetra (reactor), are too complicated to setup (sockpuppet), or are slower (Unicorn).
I really had them all...
So, it looks like I will stay with Tetra on the long term, so maybe I can help a bit out, even if my time is limited too...

About the template loader: Way do you (in `BasicComponent.get_template_source_location()`) parse the **python source** file to pull out the (Inline)Template string instead of just using Django's Template class (like `Template(cls.template)`?
I suppose you do some trickery in there, but why load the source file? `.template` is a str already...
This seems to me like printing out a PDF and scanning it again, to save that scan in a folder, instead of just saving the PDF directly :laughing:.

**Reply by samwillis** [reply]
> ... have a worse API ... complicated to setup ...

Yes, thats the driving force behind the Tetra concept, less tooling for a dev to manage. One of the next things on my To Do list is looking at replacing esbuild with Vite. It has better support for css pre/post processors, and I quite like the idea of trying to do hot module reloading eventually . The idea would be that `python manage.py runserver` would boot up both the Django and Vite dev servers. I have been using Vite recently, it's really good.

Using this other project of mine (https://pypi.org/project/nodejs-bin/) the plan is that `pip install tetraframework[all]` would also install node in your Venv. Then `python manage.py tetra init` would automatically install and configure Vite for your Django project - I think this would be useful outside of Tetra in general Django projects too, replacing existing asset packaging tools.

> So, it looks like I will stay with Tetra on the long term, so maybe I can help a bit out, even if my time is limited too...

That would be awesome!

> About the template loader: Way do you (in `BasicComponent.get_template_source_location()`) parse the **python source** file to pull out the (Inline)Template string

This is to do with Source Maps, I find the location of the template in the python file and then pad the JS/CSS source before passing it to esbuild. That way when you look at errors, css props, or the source with the browser dev tools the line numbers match those in your *.py file. Similar to how Vue does with it's single file components.

I tried a couple of ways to do this, I forget the details now, but one option was to try and locate the line the template was defined on when the class was created using the meta class... this method was far simple.

**Reply by nerdoc** [reply]
> This is to do with Source Maps, I find the location of the template in the python file and then pad the JS/CSS source before passing it to esbuild. That way when you look at errors, css props, or the source with the browser dev tools the line numbers match those in your *.py file. Similar to how Vue does with it's single file components.

OMG, therefore. Nice.

But I still am not completely convinced. Maybe just a PyCharm plugin is missing, but my simple brain relies on all the IDE helpers like reformatting, auto indenting, highlighting etc - which is poor ATM with inline templates. Vue has an advantage here, as the whole file is XML. just template/script tags are different parts, but all is the same stuff. Mixing Python with Django-HTML isn't quite easy - beginning with the basic 4x indentation on the first level.

**BadIdea™**: what if Tetra would use a template like Vue, with `<template>...</template><script>...</script><style>...</style>` - in **one** external file? So just Python code is separated. But then the usual IDE helpers would work...

**Reply by nerdoc** [reply]
FYI: I filed a [feature request](https://youtrack.jetbrains.com/issue/PY-56525/Feature-Request-Support-Tetra-components-syntax-highlighting) at PyCharm's for syntax highlighting. Don't know if they do that... I suppose that's low prio

**Reply by samwillis** [reply]
> FYI: I filed a [feature request](https://youtrack.jetbrains.com/issue/PY-56525/Feature-Request-Support-Tetra-components-syntax-highlighting) at PyCharm's for syntax highlighting. Don't know if they do that... I suppose that's low prio

Thanks, I suspect they will suggest it need to be done though an extension - I have no experience of PyCharm.

For VSCode support this is the project adding it: https://github.com/samwillis/python-inline-source

It may be possible to extend that with PyCharm support. VSCode uses TextMate Grammars, it looks like it may be possible to use them with PyCharm: https://www.jetbrains.com/help/pycharm/tutorial-using-textmate-bundles.html. However I don't know if you can inline one language in another with them in PyCharm.

**Reply by samwillis** [reply]
> **BadIdea™**: what if Tetra would use a template like Vue, with `<template>...</template><script>...</script><style>...</style>` - in **one** external file? So just Python code is separated. But then the usual IDE helpers would work...

This was actually the original idea when I started building tetra, using Python import hooks to allow you to import `*.tetra` "single file components". But when I started looking into how the Vue extension for VSCode worked I realised it was a far too large a project on its own. I then came up with the type annotation version.

**Reply by nerdoc** [reply]
Ok, thats really easier then. And I think the main argument here on my side is the IDE's helpers. So if this is solved, I really don't care if it's in another file or not. And you're right, having it together helps, AND encourages devs to create small, effective, one-purpose-doing-well components. Which is the way ;-)

**Reply by gsxdsm** [reply]
I know this is very old, but Pycharm language injection actually works really really nicely with Tetra - https://www.jetbrains.com/help/pycharm/using-language-injections.html

Both temporary injections and using #language=html, #language=css, or #language=javascript above the literals works well. You can even add a regex to look at pyStringLiterals and try to detect html ("<") or Javascript ("export default").

**Reply by nerdoc** [reply]
This is not old at all, it's still valid!
I know that in PyCharm syntax highlighting with the directives `#language=html, #language=css, or #language=javascript` are working, BUT: this is not really as helpful as it could:

1. It is lacking auto indenting support (which is ok, but not funny), PyCharm has no `#language=django` yet, so `{% if foo %}` is not detected, and all IDE helpers that are bound to special files (e.h. html) are not working in inline templates as good as in real files. etc.
2. Another issue is caching. Tetra has to do special things to allow templates to be cached.
3. whenever you change a normal template in development and reload the page, it's instantly changed.
Whenever you change an inline template, Django restarts the whole server (for each tiny change), as you in reality changed a `.py` file.
4. Very important, PyCharm doesn't allow you to do things like `button.primary[TAB]` to auto-expand to `<button class="primary"></button>` - but I think this is mainly because of 1.
5. **Django does not find translated strings within inline templates** as it searches only in .html files for it.
6. It is impossible to set breakpoints within a `django_html` template code. PyCharm only debugs .html file templates.

**Reply by nerdoc** [reply]
BTW, in  tetra > 0.1.1, you can also use a `template_name="foo.html"` instead of the inline template to use file templates.

**Reply by gsxdsm** [reply]
Great points, thanks for the reply!

**Reply by gsxdsm** [reply]
One thing I noticed - changing the template referenced with template_name= doesn't trigger a rebuild of the Tetra component bundle. Feels like it should, right?

**Reply by nerdoc** [reply]
Yes, definitely. That's because interestingly Django doesn't recognize the template as template, I still don't know why yet. I think it's about caching, and I didn't find out yet wyy this doesn't work. ATM I am working mostly on the Form field upload feature, this captures all of my spare time I have for Tetra... But the autorebuild on file templates s on my list too.
But I think opening an issue would be a good idea.
Edit: #74

**Reply by gsxdsm** [reply]
Great, I'm going to dig into this and see if I can help resolve!

**Reply by nerdoc** [reply]
So, happily this is completed in 0.1.2, with many other improvements (and some breaking changes, watch out!)
